extensions/ffi/lib.rs

```diff
@@ -0,0 +1,329 @@
+    match native_type {
+      NativeType::Void => libffi::middle::Type::void(),
+      NativeType::U8 => libffi::middle::Type::u8(),
+      NativeType::I8 => libffi::middle::Type::i8(),
+      NativeType::U16 => libffi::middle::Type::u16(),
+      NativeType::I16 => libffi::middle::Type::i16(),
+      NativeType::U32 => libffi::middle::Type::u32(),
+      NativeType::I32 => libffi::middle::Type::i32(),
+      NativeType::U64 => libffi::middle::Type::u64(),
+      NativeType::I64 => libffi::middle::Type::i64(),
+      NativeType::USize => libffi::middle::Type::usize(),
+      NativeType::ISize => libffi::middle::Type::isize(),
+      NativeType::F32 => libffi::middle::Type::f32(),
+      NativeType::F64 => libffi::middle::Type::f64(),
+    }
+  }
+}
+
+impl From<String> for NativeType {
+  fn from(string: String) -> Self {
+    match string.as_str() {
+      "void" => NativeType::Void,
+      "u8" => NativeType::U8,
+      "i8" => NativeType::I8,
+      "u16" => NativeType::U16,
+      "i16" => NativeType::I16,
+      "u32" => NativeType::U32,
+      "i32" => NativeType::I32,
+      "u64" => NativeType::U64,
+      "i64" => NativeType::I64,
+      "usize" => NativeType::USize,
+      "isize" => NativeType::ISize,
+      "f32" => NativeType::F32,
+      "f64" => NativeType::F64,
+      _ => unimplemented!(),
+    }
+  }
+}
+
+#[derive(Deserialize)]
+struct ForeignFunction {
+  parameters: Vec<String>,
+  result: String,
+}
+
+#[derive(Deserialize)]
+struct FfiLoadArgs {
+  path: String,
+  symbols: HashMap<String, ForeignFunction>,
+}
+
+fn op_ffi_load<FP>(
+  state: &mut deno_core::OpState,
+  args: FfiLoadArgs,
+  _: (),
+) -> Result<ResourceId, AnyError>
+where
+  FP: FfiPermissions + 'static,
+{
+  check_unstable(state, "Deno.dlopen");
+  let permissions = state.borrow_mut::<FP>();
+  permissions.check(&args.path)?;
+
+  let lib = Library::open(args.path)?;
+  let mut resource = DynamicLibraryResource {
+    lib,
+    symbols: HashMap::new(),
+  };
+
+  for (symbol, foreign_fn) in args.symbols {
+    resource.register(symbol, foreign_fn)?;
+  }
+
+  Ok(state.resource_table.add(resource))
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct FfiCallArgs {
+  rid: ResourceId,
+  symbol: String,
+  parameters: Vec<Value>,
+}
+
+fn op_ffi_call(
+  state: &mut deno_core::OpState,
+  args: FfiCallArgs,
+  _: (),
+) -> Result<Value, AnyError> {
+  let resource = state
+    .resource_table
+    .get::<DynamicLibraryResource>(args.rid)
+    .ok_or_else(bad_resource_id)?;
+
+  let symbol = resource
+    .symbols
+    .get(&args.symbol)
+    .ok_or_else(bad_resource_id)?;
+
+  let call_args: Vec<libffi::middle::Arg> = symbol
+    .parameter_types
+    .iter()
+    .zip(args.parameters.into_iter())
+    .map(|(parameter_type, value)| match parameter_type {
+      NativeType::Void => libffi::middle::Arg::new(&()),
+      NativeType::U8 => libffi::middle::Arg::new(&(value_as_u64(value) as u8)),
```

Comment:
Passing a reference to an on-stack variable to `Arg::new()` is invalid (and unsafe). 
This is a libffi-rs bug btw.